foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/EnableHabushuBuildCacheMigration.java

```diff
@@ -0,0 +1,112 @@
+            "<xsl:stylesheet version=\"1.0\" xmlns:xsl=\"http://www.w3.org/1999/XSL/Transform\">\n" +
+            "  <xsl:output indent=\"yes\"/>\n" +
+            "  <xsl:strip-space elements=\"*\"/>\n" +
+            "\n" +
+            "  <xsl:template match=\"@*|node()\">\n" +
+            "    <xsl:copy>\n" +
+            "      <xsl:apply-templates select=\"@*|node()\"/>\n" +
+            "    </xsl:copy>\n" +
+            "  </xsl:template>\n" +
+            "\n" +
+            "</xsl:stylesheet>";
+
+    /**
+     * Retrieves the xml node at the specified path
+     *
+     * @param file The xml file to parse
+     * @param expression The path to the target node
+     * @return The retrieved node, or null if the node at the specified path does not exist.
+     */
+    public static Node getNode(File file, String expression) {
+        DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder builder;
+        try {
+            builder = builderFactory.newDocumentBuilder();
+            Document xmlDocument = builder.parse(file);
+            XPath xPath = XPathFactory.newInstance().newXPath();
+            return (Node) xPath.compile(expression).evaluate(xmlDocument, XPathConstants.NODE);
+        } catch (ParserConfigurationException | XPathExpressionException | IOException | SAXException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean isHabushuPackaged(File file) {
+        Node node = getNode(file, PACKAGING_NODE_PATH);
+        return node != null && node.getTextContent().equalsIgnoreCase("habushu");
+    }
+
+    private boolean doesBuildDirTagExist(File file) {
+        Node buildDirNode = getNode(file, BUILD_DIR_NODE_PATH);
+        return buildDirNode != null;
+    }
+
+    @Override
+    protected boolean shouldExecuteOnFile(File file) {
+        return file != null && file.exists() && !doesBuildDirTagExist(file) && isHabushuPackaged(file);
+    }
+
+    @Override
+    protected boolean performMigration(File file) {
+        // Add the new child node...
+        Node buildNode = getNode(file, BUILD_NODE_PATH);
+        Node dirNode = buildNode.appendChild(buildNode.getOwnerDocument().createElement("directory"));
+        dirNode.setTextContent("dist");
+
+        // ...and write it back to the filesystem.
+        TransformerFactory transformerFactory = TransformerFactory.newInstance();
+        try {
+            Transformer transformer = transformerFactory.newTransformer(new StreamSource(new StringReader(xsl)));
+            transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+            transformerFactory.setAttribute("indent-number", 4);
+            DOMSource source = new DOMSource(dirNode.getOwnerDocument());
+            FileWriter writer = new FileWriter(file);
+            StreamResult result = new StreamResult(writer);
+            transformer.transform(source, result);
+            writer.close();
```

Comment:
S: We usually try not to write out the new file from its java representation as you could potentially lose any formatting/comments. Does this method preserve and comments that may exist in the `pom.xml` prior to the migration?

Reply:
Yes it does. Comments are a valid type of node.